src/app_conn_pool.rs

```diff
@@ -0,0 +1,319 @@
+            ))
+            .await?;
+
+        // Presence of host must have been checked to get an admin connection
+        let host = self
+            .configuration
+            .admin_ws_url
+            .host_str()
+            .expect("Must have a host");
+
+        tracing::info!("Connecting to app websocket at {}:{}", host, app_port);
+
+        // Build a connection request
+        let request =
+            ConnectRequest::from(format!("{host}:{app_port}").parse::<SocketAddr>().map_err(
+                |e| {
+                    HcHttpGatewayError::ConfigurationError(format!("Invalid socket address: {}", e))
+                },
+            )?)
+            .try_set_header("Origin", HTTP_GW_ORIGIN)?;
+
+        // Create a websocket client configuration and lower the default timeout. We are connecting
+        // locally to a running Holochain. If requests take longer than 10s then we want to free up
+        // the HTTP gateway to handle other requests.
+        let mut config = WebsocketConfig::CLIENT_DEFAULT;
+        config.default_request_timeout = std::time::Duration::from_secs(10);
+
+        let client_signer = ClientAgentSigner::default();
+
+        // Attempt to connect to the app websocket
+        let app_ws = match AppWebsocket::connect_with_request_and_config(
+            request,
+            Arc::new(config),
+            issued.token,
+            client_signer.clone().into(),
+        )
+        .await
+        {
+            Ok(client) => client,
+            Err(e) => {
+                tracing::error!("Failed to connect to app websocket: {}", e);
+
+                // If we failed to make a connection, clear the cached app port so that the next
+                // attempt will re-check the app interfaces.
+                *self.cached_app_port.write().expect("Invalid lock") = None;
+
+                // Try again, with one fewer retry permitted
+                return Box::pin(self.attempt_connect_app_ws(
+                    installed_app_id,
+                    admin_ws,
+                    retries - 1,
+                ))
+                .await;
+            }
+        };
+        tracing::debug!("Connected to app websocket");
+
+        let app_info = app_ws.cached_app_info();
+        let cells = app_info
+            .cell_info
+            .values()
+            .flat_map(|cell_infos| {
+                cell_infos.iter().flat_map(|cell_info| {
+                    match cell_info {
+                        CellInfo::Provisioned(provisioned) => Some(provisioned.cell_id.clone()),
+                        // TODO Provisioning of these wouldn't be dynamic, you'd have to
+                        //      restart the gateway or Holochain to get new credentials for
+                        //      new clones...
+                        //      See https://github.com/holochain/holochain/issues/4595
+                        // CellInfo::Cloned(clone_cell) => Some(clone_cell.cell_id.clone()),
```

Comment:
The way this is written, we still wouldn't get credentials for clones, even after restarting right?

Reply:
I thought the same yes. The way it is right now, cloned cells are not supported categorically.
I'm also noticing that granting of functions is only by zome name and function name and not role name. So there would be no way currently to restrict calling a function for specific clones or roles only.

Reply:
Oh I see what you're saying, yes that's a good point. So you could either permit access to all clones if the DNA hash is known or you'd have to add more configuration to the gateway to restrict which clones are permitted.